smithy-typescript-codegen/src/main/java/software/amazon/smithy/typescript/codegen/integration/DocumentShapeSerVisitor.java

```diff
@@ -0,0 +1,440 @@
+import software.amazon.smithy.codegen.core.Symbol;
+import software.amazon.smithy.codegen.core.SymbolProvider;
+import software.amazon.smithy.model.shapes.CollectionShape;
+import software.amazon.smithy.model.shapes.DocumentShape;
+import software.amazon.smithy.model.shapes.ListShape;
+import software.amazon.smithy.model.shapes.MapShape;
+import software.amazon.smithy.model.shapes.OperationShape;
+import software.amazon.smithy.model.shapes.ResourceShape;
+import software.amazon.smithy.model.shapes.ServiceShape;
+import software.amazon.smithy.model.shapes.SetShape;
+import software.amazon.smithy.model.shapes.Shape;
+import software.amazon.smithy.model.shapes.ShapeVisitor;
+import software.amazon.smithy.model.shapes.StructureShape;
+import software.amazon.smithy.model.shapes.UnionShape;
+import software.amazon.smithy.typescript.codegen.TypeScriptWriter;
+import software.amazon.smithy.typescript.codegen.integration.ProtocolGenerator.GenerationContext;
+
+/**
+ * Visitor to generate serialization functions for shapes in protocol document bodies.
+ *
+ * Visitor methods for aggregate types are final and will generate functions that dispatch
+ * their body generation to the matching abstract method. The {@link DocumentMemberSerVisitor}
+ * is provided to reduce the effort of this implementation by providing the default strategies
+ * for serializing content from aggregate type members.
+ *
+ * Visitor methods for all other types will default to not generating serialization functions.
+ * This may be overwritten by downstream implementations if the protocol requires more
+ * complex serialization strategies for those types.
+ */
+public abstract class DocumentShapeSerVisitor extends ShapeVisitor.Default<Void> {
+    private GenerationContext context;
+    private String protocolName;
+
+    public DocumentShapeSerVisitor(GenerationContext context, String protocolName) {
+        this.context = context;
+        this.protocolName = protocolName;
+    }
+
+    /**
+     * Gets the generation context.
+     *
+     * @return The generation context.
+     */
+    protected GenerationContext getContext() {
+        return context;
+    }
+
+    /**
+     * Gets the name of the protocol.
+     *
+     * @return The protocol name.
+     */
+    protected String getProtocolName() {
+        return protocolName;
+    }
+
+    @Override
+    protected Void getDefault(Shape shape) {
+        return null;
+    }
+
+    /**
+     * Writes the code needed to serialize a collection in the document of a request.
+     *
+     * <p>Implementations of this method are expected to generate a function body that
+     * returns a value representing the CollectionShape {@code shape} parameter that is
+     * serializable by {@code serializeInputDocument}.
+     *
+     * <p>The function signature for this body will have two parameters available in scope:
+     * <ul>
+     *   <li>{@code input}: the type generated for the CollectionShape shape parameter.</li>
```

Comment:
Again from other comment -- these should be more specific about the TypeScript types of each of these variables